### ISSUE TITLE
parser configuration isn't very usable in UI

### ISSUE THREAD

**Issue by jgraettinger** [estuary/connectors#221]
I've [made some tweaks](https://github.com/estuary/flow/commit/73dd54bbc657b7275f0b8998a43ea621c885a2b5) to the parser config which allow it to now render in the UI. You can see it w/ the S3 & GCS connectors. I'm having to put this down for now, but reporting some notes of remaining frictions where I'm leaving off, as it requires deeper changes than I can commit to at the moment.

* The Option<Format> type would probably work better as a non-Option with an Infer variant, as the user can't otherwise de-select a setting for this enum.

* Delimiter is also hard to use, since you can't enter \n or \t in the UI. Line separator has the same issue.

* Things like extra values, headers, file extension mappings, and content-type mappings should probably be hidden or even removed since there's no reasonable way for a user to modify these in the UI.

* Having seen it in the UI, I think I'd still push for consolidating CSV vs TSV configuration, as it does feel confusing to me. We could instead assume , vs \t from the file extension if the delimiter isn't set.

**Comment by psFried** [comment]
After some back and forth with @travjenkins , we now have a desired parser schema that is known to work in the UI. The example schema is below. At this point, the remaining work is just to update the parser schema to match this. This requires the use of the [discriminator annotation](https://spec.openapis.org/oas/latest.html#discriminator-object) from openapi. Ajv (and therefore jsonForms) supports this annotation to help drive the behavior of `oneOf` controls.

<details>
<summary>Example JSON schema</summary>

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "GCS Source",
  "type": "object",
  "required": ["bucket"],
  "properties": {
    "ascendingKeys": {
      "title": "Ascending Keys",
      "description": "Improve sync speeds by listing files from the end of the last sync, rather than listing the entire bucket prefix. This requires that you write objects in ascending lexicographic order, such as an RFC-3339 timestamp, so that key ordering matches modification time ordering.",
      "default": false,
      "type": "boolean"
    },
    "bucket": {
      "title": "Bucket",
      "description": "Name of the Google Cloud Storage bucket",
      "type": "string"
    },
    "googleCredentials": {
      "title": "Google Service Account",
      "description": "Service account JSON key to use as Application Default Credentials",
      "type": "string",
      "multiline": true,
      "secret": true
    },
    "matchKeys": {
      "title": "Match Keys",
      "description": "Filter applied to all object keys under the prefix. If provided, only objects whose key (relative to the prefix) matches this regex will be read. For example, you can use \".*\\.json\" to only capture json files.",
      "type": "string",
      "format": "regex"
    },
    "networkTunnel": {
      "title": "Network Tunneling",
      "description": "Setup a network tunnel to access systems on a private network",
      "oneOf": [
        {
          "title": "SSH Forwarding",
          "type": "object",
          "properties": {
            "sshForwarding": {
              "title": "SSH Tunnel",
              "description": "Connect to your system through an SSH server that acts as a bastion host for your network.",
              "type": "object",
              "required": ["privateKey", "sshEndpoint"],
              "properties": {
                "forwardHost": {
                  "description": "The hostname of the remote destination (e.g. the database server).",
                  "default": "",
                  "type": "string"
                },
                "forwardPort": {
                  "title": "Forward Port",
                  "description": "The port of the remote destination (e.g. the database server).",
                  "default": 0,
                  "type": "integer",
                  "maximum": 65535,
                  "minimum": 1
                },
                "localPort": {
                  "title": "Local Port",
                  "description": "The local port which will be connected to the remote host/port over an SSH tunnel. This should match the port that's used in your basic connector configuration.",
                  "default": 0,
                  "type": "integer",
                  "maximum": 65535,
                  "minimum": 1
                },
                "privateKey": {
                  "title": "SSH Private Key",
                  "description": "Private key to connect to the remote SSH server.",
                  "type": "string",
                  "multiline": true,
                  "secret": true
                },
                "sshEndpoint": {
                  "description": "Endpoint of the remote SSH server that supports tunneling, in the form of ssh://user@hostname[:port]",
                  "type": "string"
                }
              }
            }
          }
        },
        {
          "title": "Disabled",
          "type": "null"
        }
      ],
      "advanced": true
    },
    "parser": {
      "$id": "uuid://789bdc9c-e4a0-4020-a0b6-fb75026dcb04",
      "title": "Parser Configuration",
      "description": "Configures how files are parsed",
      "type": "object",
      "properties": {
        "compression": {
          "title": "Compression",
          "description": "Determines how to decompress the contents. The default, 'Auto', will try to determine the compression automatically.",
          "default": null,
          "oneOf": [
            {
              "title": "GZip",
              "const": "gzip"
            },
            {
              "title": "Zip Archive",
              "const": "zip"
            },
            {
              "title": "None",
              "const": "none"
            },
            {
              "title": "Auto",
              "const": null
            }
          ]
        },
        "format": {
          "description": "Determines how to parse the contents. The default, 'Auto', will try to determine the format automatically based on the file extension or MIME type, if available.",
          "default": {
            "formatter": "auto"
          },
          "discriminator": {
            "propertyName": "formatter"
          },
          "required": ["formatter"],
          "oneOf": [
            {
              "title": "Auto",
              "description": "Attempt to determine the format automatically, based on the file extension or associated content-type.",
              "type": "object",
              "required": ["formatter"],
              "properties": {
                "formatter": {
                  "const": "auto",
                  "default": "auto"
                }
              },
              "additionalProperties": false
            },
            {
              "title": "JSON",
              "description": "JSON objects separated by whitespace, typically a single newline. This format works for JSONL (a.k.a. JSON-newline), but also for any stream of JSON objects, as long as they have at least one character of whitespace between them.",
              "type": "object",
              "required": ["formatter"],
              "properties": {
                "formatter": {
                  "const": "json",
                  "default": "json"
                }
              },
              "additionalProperties": false
            },
            {
              "title": "Avro",
              "description": "Avro object container files, as defined by the [avro spec](https://avro.apache.org/docs/current/spec.html#Object+Container+Files)",
              "type": "object",
              "required": ["formatter"],
              "properties": {
                "formatter": {
                  "const": "avro",
                  "default": "avro"
                }
              },
              "additionalProperties": false
            },
            {
              "title": "CSV",
              "description": "Character Separated Values, such as comma-separated, tab-separated, etc.",
              "type": "object",
              "required": ["formatter", "settings"],
              "properties": {
                "formatter": {
                  "const": "csv",
                  "default": "csv"
                },
                "settings": {
                  "type": "object",
                  "properties": {
                    "delimiter": {
                      "title": "Delimiter",
                      "description": "The delimiter that separates values within each row. Only single-byte delimiters are supported.",
                      "default": null,
                      "oneOf": [
                        {
                          "title": "Comma (,)",
                          "const": ","
                        },
                        {
                          "title": "Pipe (|)",
                          "const": "|"
                        },
                        {
                          "title": "Space (0x20)",
                          "const": " "
                        },
                        {
                          "title": "Semicolon (;)",
                          "const": ";"
                        },
                        {
                          "title": "Tab (0x09)",
                          "const": "\t"
                        },
                        {
                          "title": "Vertical Tab (0x0B)",
                          "const": "\u000b"
                        },
                        {
                          "title": "Unit Separator (0x1F)",
                          "const": "\u001f"
                        },
                        {
                          "title": "SOH (0x01)",
                          "const": "\u0001"
                        },
                        {
                          "title": "Auto",
                          "const": null
                        }
                      ]
                    },
                    "encoding": {
                      "title": "Encoding",
                      "description": "The character encoding of the source file. If unspecified, then the parser will make a best-effort guess based on peeking at a small portion of the beginning of the file. If known, it is best to specify. Encodings are specified by their WHATWG label.",
                      "default": null,
                      "oneOf": [
                        {
                          "title": "UTF-8",
                          "const": "UTF-8"
                        },
                        {
                          "title": "UTF-16LE",
                          "const": "UTF-16LE"
                        },
                        {
                          "title": "UTF-16BE",
                          "const": "UTF-16BE"
                        },
                        {
                          "title": "IBM866",
                          "const": "IBM866"
                        },
                        {
                          "title": "ISO-8859-2",
                          "const": "ISO-8859-2"
                        },
                        {
                          "title": "ISO-8859-3",
                          "const": "ISO-8859-3"
                        },
                        {
                          "title": "ISO-8859-4",
                          "const": "ISO-8859-4"
                        },
                        {
                          "title": "ISO-8859-5",
                          "const": "ISO-8859-5"
                        },
                        {
                          "title": "ISO-8859-6",
                          "const": "ISO-8859-6"
                        },
                        {
                          "title": "ISO-8859-7",
                          "const": "ISO-8859-7"
                        },
                        {
                          "title": "ISO-8859-8",
                          "const": "ISO-8859-8"
                        },
                        {
                          "title": "ISO-8859-8-I",
                          "const": "ISO-8859-8-I"
                        },
                        {
                          "title": "ISO-8859-10",
                          "const": "ISO-8859-10"
                        },
                        {
                          "title": "ISO-8859-13",
                          "const": "ISO-8859-13"
                        },
                        {
                          "title": "ISO-8859-14",
                          "const": "ISO-8859-14"
                        },
                        {
                          "title": "ISO-8859-15",
                          "const": "ISO-8859-15"
                        },
                        {
                          "title": "ISO-8859-16",
                          "const": "ISO-8859-16"
                        },
                        {
                          "title": "KOI8-R",
                          "const": "KOI8-R"
                        },
                        {
                          "title": "KOI8-U",
                          "const": "KOI8-U"
                        },
                        {
                          "title": "macintosh",
                          "const": "macintosh"
                        },
                        {
                          "title": "windows-874",
                          "const": "windows-874"
                        },
                        {
                          "title": "windows-1250",
                          "const": "windows-1250"
                        },
                        {
                          "title": "windows-1251",
                          "const": "windows-1251"
                        },
                        {
                          "title": "windows-1252",
                          "const": "windows-1252"
                        },
                        {
                          "title": "windows-1253",
                          "const": "windows-1253"
                        },
                        {
                          "title": "windows-1254",
                          "const": "windows-1254"
                        },
                        {
                          "title": "windows-1255",
                          "const": "windows-1255"
                        },
                        {
                          "title": "windows-1256",
                          "const": "windows-1256"
                        },
                        {
                          "title": "windows-1257",
                          "const": "windows-1257"
                        },
                        {
                          "title": "windows-1258",
                          "const": "windows-1258"
                        },
                        {
                          "title": "x-mac-cyrillic",
                          "const": "x-mac-cyrillic"
                        },
                        {
                          "title": "GBK",
                          "const": "GBK"
                        },
                        {
                          "title": "gb18030",
                          "const": "gb18030"
                        },
                        {
                          "title": "Big5",
                          "const": "Big5"
                        },
                        {
                          "title": "EUC-JP",
                          "const": "EUC-JP"
                        },
                        {
                          "title": "ISO-2022-JP",
                          "const": "ISO-2022-JP"
                        },
                        {
                          "title": "Shift_JIS",
                          "const": "Shift_JIS"
                        },
                        {
                          "title": "EUC-KR",
                          "const": "EUC-KR"
                        },
                        {
                          "title": "Auto",
                          "const": null
                        }
                      ]
                    },
                    "errorThreshold": {
                      "title": "Error Threshold",
                      "description": "Allows a percentage of errors to be ignored without failing the entire parsing process. When this limit is exceeded, parsing halts.",
                      "default": 1,
                      "type": "integer",
                      "maximum": 100,
                      "minimum": 0
                    },
                    "escape": {
                      "title": "Escape Character",
                      "description": "The escape character, used to escape quotes within fields.",
                      "default": null,
                      "oneOf": [
                        {
                          "title": "Backslash (\\)",
                          "const": "\\"
                        },
                        {
                          "title": "Disable Escapes",
                          "const": ""
                        },
                        {
                          "title": "Auto",
                          "const": null
                        }
                      ]
                    },
                    "headers": {
                      "description": "Manually specified headers, which can be used in cases where the file itself doesn't contain a header row. If specified, then the parser will assume that the first row is data, not column names, and the column names given here will be used. The column names will be matched with the columns in the file by the order in which they appear here.",
                      "default": [],
                      "type": "array",
                      "items": {
                        "type": "string"
                      }
                    },
                    "lineEnding": {
                      "title": "Line Ending",
                      "description": "The value that terminates a line. Only single-byte values are supported, withe the exception of \"\\r\\n\" (CRLF), which will accept lines terminated by _either_ a carriage return, a newline, or both.",
                      "default": null,
                      "oneOf": [
                        {
                          "title": "CRLF (\\r\\n) (Windows)",
                          "const": "\r\n"
                        },
                        {
                          "title": "CR (\\r)",
                          "const": "\r"
                        },
                        {
                          "title": "LF (\\n)",
                          "const": "\n"
                        },
                        {
                          "title": "Record Separator (0x1E)",
                          "const": "\u001e"
                        },
                        {
                          "title": "Auto",
                          "const": null
                        }
                      ]
                    },
                    "quote": {
                      "title": "Quote Character",
                      "description": "The character used to quote fields.",
                      "default": null,
                      "oneOf": [
                        {
                          "title": "Double Quote (\")",
                          "const": "\""
                        },
                        {
                          "title": "Single Quote (')",
                          "const": "'"
                        },
                        {
                          "title": "Disable Quoting",
                          "const": ""
                        },
                        {
                          "title": "Auto",
                          "const": null
                        }
                      ]
                    }
                  }
                }
              },
              "additionalProperties": false
            },
            {
              "title": "W3C Extended Log",
              "description": "A W3C Extended Log file, as defined by the working group draft at: https://www.w3.org/TR/WD-logfile.html",
              "type": "object",
              "required": ["formatter"],
              "properties": {
                "formatter": {
                  "const": "w3cExtendedLog",
                  "default": "w3cExtendedLog"
                }
              },
              "additionalProperties": false
            }
          ]
        }
      },
      "$schema": "http://json-schema.org/draft-07/schema#"
    },
    "prefix": {
      "title": "Prefix",
      "description": "Prefix within the bucket to capture from",
      "type": "string"
    }
  }
}
```
</details>

**Comment by psFried** [comment]
The remainder of the issues have now been resolved by estuary/flow#591
